### PR TITLE
Allow @implicitNotFound messages as explanations

### PR DESCRIPTION
**Pull request by odersky**
A problem of the @implicitNotFOund mechanism so far was that the user defined message replaced the compiler-generated one, which might lose valuable information.

This commit adds an alternative where an @implicitNotFound message that starts with `...` is taken as an explanation (without the ...) enabled under -explain. The compiler-generated message is then kept as the explicit error message.

We apply the mechanism for an @implicitNotFound message for `boundary.Label`. This now produces messages like this one:
```
-- [E172] Type Error: tests/neg-custom-args/explain/labelNotFound.scala:2:30 -------------------------------------------
2 |  scala.util.boundary.break(1) // error
  |                              ^
  |No given instance of type scala.util.boundary.Label[Int] was found for parameter label of method break in object boundary
  |---------------------------------------------------------------------------------------------------------------------
  | Explanation (enabled by `-explain`)
  |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  | A Label is generated from an enclosing `scala.util.boundary` call.
  | Maybe that boundary is missing?
   ---------------------------------------------------------------------------------------------------------------------
```

### CODE REVIEW

**Comment by odersky** [comment]
@nicolasstucki We can use an analogous mechanism to improve the situation for #16888.

**Comment by prolativ** [comment]
Should we somehow document this new behaviour or is it supposed to be used only internally?

**Comment by nicolasstucki** [comment]
Unfortunately, the spec of `implicitNotFound` states how the error will be shown. Changing this could break existing use cases.

A solution to this might be to show the custom message and hide the compiler-generated message under `-explain`.

The message ``A Label is generated from an enclosing `scala.util.boundary` call. Maybe that boundary is missing?`` does not seem to fit exactly the intended use of `implicitNotFound`. It feels like we should have a way to customize the message and the `-explain`. The example would be the custom `-explain`. We could modify the `class implicitNotFound` to take a message and explain string or add a separate `@implicitNotFoundExplain`.

Maybe this is how an error message should look
```
-- [E172] Type Error: tests/neg-custom-args/explain/labelNotFound.scala:2:30 -------------------------------------------
2 |  scala.util.boundary.break(1) // error
  |                              ^
  |  missing Int boundary label
  |---------------------------------------------------------------------------------------------------------------------
  | Explanation (enabled by `-explain`)
  |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  | No given instance of type scala.util.boundary.Label[Int] was found for parameter label of method break in object boundary
  |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  | A Label[Int] is generated from an enclosing `scala.util.boundary` call.
  | Maybe that boundary is missing?
  | Maybe a given Label[Int] is missing from the parameters?
   ---------------------------------------------------------------------------------------------------------------------
```

**Comment by odersky** [comment]
> Unfortunately, the spec of implicitNotFound states how the error will be shown. Changing this could break existing use cases.

In theory yes, in practice, I would be _really_ surprised if someone started an `implicitNotFound` message with `...`.

**Comment by odersky** [comment]
But I guess it's OK to add an `implicitNoFoundExplain` annotation. Do you want to add that one? I just wanted to show what needs to be done, but won't have the time to do it.

**Comment by prolativ** [comment]
Adding a new annotation would have to be done in a minor release, unless we make it experimental

**Comment by bishabosha** [comment]
> Adding a new annotation would have to be done in a minor release, unless we make it experimental

the release candidates still have experimental tasty version, so its ok?

**Comment by odersky** [comment]
So somebody should do it quickly! These are real and very important usability improvements. We should not block them by procedural reasons.

**Comment by prolativ** [comment]
How would we like this new annotation to be used then? Should it be orthogonal to `implicitNotFound`, meaning that the old annotation would only replace the error message and the new one would only add explanation?

**Comment by odersky** [comment]
> How would we like this new annotation to be used then? Should it be orthogonal to implicitNotFound, meaning that the old annotation would only replace the error message and the new one would only add explanation?

Yes, I think that's reasonable.

**Comment by smarter** [comment]
In Scala 2, `@implicitNotFound` does not replace the message, both are shown. This difference in behavior has lead to some confusion as described in https://github.com/lampepfl/dotty/issues/11493, and it seems people are happy with doing it the Scala 2 way which wouldn't require introducing a new annotation.

**Comment by odersky** [comment]
@smarter I did not know that Scala 2 behaved differently. Yes, let's go back to that then the whole discussion about explain becomes redundant.

**Comment by nicolasstucki** [comment]
> In Scala 2, @implicitNotFound does not replace the message, both are shown. This difference in behavior has lead to some confusion as described in https://github.com/lampepfl/dotty/issues/11493, and it seems people are happy with doing it the Scala 2 way which wouldn't require introducing a new annotation.

It does not seem to be the case. In Scala 2 only the custom message is shown. It seems that the behaviour is the same as in Scala 3.

```scala
@scala.annotation.implicitNotFound("Foo[${T}] was not found")
final class Foo[T]
class C {
  def test = implicitly[Foo[Int]]
}
```

```
~/GitHub/dotty @ main % scala-cli compile -S 2.13 Foo.scala
Compiling project (Scala 2.13.10, JVM)
[error] ./Foo.scala:4:14: Foo[Int] was not found
[error]   def test = implicitly[Foo[Int]]
[error]              ^^^^^^^^^^^^^^^^^^^^
Error compiling project (Scala 2.13.10, JVM)
Compilation failed
~/GitHub/dotty @ main % scala-cli compile -S 3.2 Foo.scala
Compiling project (Scala 3.2.2, JVM)
[error] ./Foo.scala:4:34: Foo[Int] was not found
[error]   def test = implicitly[Foo[Int]]
[error]                                  ^
Error compiling project (Scala 3.2.2, JVM)
Compilation failed
```

**Comment by prolativ** [comment]
It looks like the behaviour in scala 2 depends on whether the annotation is inherited or not.
```scala
import annotation.implicitNotFound

@implicitNotFound("Missing type: Foo")
trait Foo

@implicitNotFound("Missing type: Bar")
trait Bar

trait Baz extends Bar

object Test {
  def myImplicitly[T](implicit @implicitNotFound("Missing parameter: ${T}") t: T): T = t

  val a = implicitly[Int]
  val b = implicitly[Foo]
  val c = implicitly[Baz]

  val x = myImplicitly[Int]
  val y = myImplicitly[Foo]
  val z = myImplicitly[Baz]
}
```
Compiling with scala 2:
```shell
scala-cli compile Test.scala -S 2.13.10
```
```scala
[error] ./Test.scala:14:11: could not find implicit value for parameter e: Int
[error]   val a = implicitly[Int]
[error]           ^^^^^^^^^^^^^^^
[error] ./Test.scala:15:11: Missing type: Foo
[error]   val b = implicitly[Foo]
[error]           ^^^^^^^^^^^^^^^
[error] ./Test.scala:16:11: could not find implicit value for parameter e: Baz (Missing type: Bar)
[error]   val c = implicitly[Baz]
[error]           ^^^^^^^^^^^^^^^
[error] ./Test.scala:18:11: Missing parameter: Int
[error]   val x = myImplicitly[Int]
[error]           ^^^^^^^^^^^^^^^^^
[error] ./Test.scala:19:11: Missing parameter: Foo
[error]   val y = myImplicitly[Foo]
[error]           ^^^^^^^^^^^^^^^^^
[error] ./Test.scala:20:11: Missing parameter: Baz
[error]   val z = myImplicitly[Baz]
[error]           ^^^^^^^^^^^^^^^^^
```
Compiling with scala 3:
```shell
scala-cli compile Test.scala -S 3.3.0-RC2
```
```scala
[error] ./Test.scala:14:26: No given instance of type Int was found for parameter e of method implicitly in object Predef
[error]   val a = implicitly[Int]
[error]                          ^
[error] ./Test.scala:15:26: Missing type: Foo
[error]   val b = implicitly[Foo]
[error]                          ^
[error] ./Test.scala:16:26: Missing type: Bar
[error]   val c = implicitly[Baz]
[error]                          ^
[error] ./Test.scala:18:28: Missing parameter: Int
[error]   val x = myImplicitly[Int]
[error]                            ^
[error] ./Test.scala:19:28: Missing parameter: Foo
[error]   val y = myImplicitly[Foo]
[error]                            ^
[error] ./Test.scala:20:28: Missing parameter: Baz
[error]   val z = myImplicitly[Baz]
[error]                            ^
```

**Comment by odersky** [comment]
I'd vote for showing both messages always. It seems the implicit not found messages are not always that specific, so adding info which implicit parameter was not found is useful.

**Comment by nicolasstucki** [comment]
> I'd vote for showing both messages always. It seems the implicit not found messages are not always that specific, so adding info which implicit parameter was not found is useful.

Sounds good. But we should put the custom error first.

**Comment by odersky** [comment]
> Sounds good. But we should put the custom error first.

I am not sure about that. Can we do play out some scenarios how it would look?

For the missing label we'd get with custom error _last_:
```
No given instance of type scala.util.boundary.Label[Int] was found for parameter label of method break in object boundary

A Label is generated from an enclosing `scala.util.boundary` call.
Maybe that boundary is missing?`
```
With custom error first this makes no sense.

Another use case is the implicitNotFound for `<:<` Here we would have with custom error first:

```
Cannot prove that Int <:< String.

No given instance of type <:<[Int, String] was found for parameter xs of method flatten.
```
Or, custom error last:
```
No given instance of type <:<[Int, String] was found for parameter xs of method flatten.

Cannot prove that Int <:< String.
```
Here it's debatable, but I think in the end I would also prefer custom error last.

**Comment by prolativ** [comment]
I believe we have two kinds of use cases:
1. The `@implicitNotFound` annotation is intended to hide the fact that some library uses implicits in it's API to make it more friendly to users who are not familiar with the concept of implicits, e.g.
```scala
package calendar

import annotation.implicitNotFound

trait DateFormat

object DateFormats {
  implicit object DayMonthYear extends DateFormat
  implicit object MonthDayYear extends DateFormat
}

class Date(day: Int, month: Int, year: Int) {
  def show(implicit @implicitNotFound("To show a date one needs to specify a date format, e.g. with `import calendar.DateFormats.DayMonthYear`") dateFormat: DateFormat): String = ???
}
```
In this situation it might make sense to:
a) show only the custom message by default
b) show the custom message followed by the generic `No given instance...` message and a link to the documentation describing what implicits/givens are, when `-explain` is enabled

2. The message from `@implicitNotFound` is intended to give some clue (to slightly more advanced users) why exactly a given instance was not found - in this case the custom message seems to make more sense when shown after the generic message

The problem is that it's not so easy to tell which situation we're dealing with in general

**Comment by prolativ** [comment]
In general it seems like `@implicitNotFound` is slightly controversial, e.g. https://github.com/typelevel/cats/pull/4060

**Comment by odersky** [comment]
I believe the cats example shows that we want _additional_ info, not a replacement of the error message. E.g. see this comment:
> I think this annotation is still useful, if you want to provide additional context (e.g., how to get that implicit). That seems to have been the original goal in https://github.com/typelevel/cats/issues/581, although it was never executed (?) and isn't relevant anymore anyway. But just saying what is missing is something the compiler does anyway, and I think (hope) has been doing for a while? :)

**Comment by prolativ** [comment]
Yeah, I would say cats classifies as case 2

**Comment by odersky** [comment]
And I would argue the DateFormat case also classifies as case 2, not 1. Taken alone, the implicitNotFound message is more confusing than helpful. If it is shown after the regular message it makes sense. 

I think by now we have not seen a single case where custom first is clearly better and lots of cases where it makes not sense at all. Let's close this issue and implement it.

**Comment by nicolasstucki** [comment]
I ended up keeping the current scheme for generating `-explain` messages but modified the format slightly to make it clearer and easily extensible. See https://github.com/lampepfl/dotty/pull/16893/commits/7de74bd2a2689af1a33fabc9bdd5e58d5f5d7668.

I also added the implicit not found details for `Quotes`.

**Review by nicolasstucki — APPROVED** [review]

**Comment by nicolasstucki** [comment]
@sjrd could you review 7de74bd? The spec part of it.

**Comment by sjrd** [comment]
What do you call "the spec part of it"? The commit message? I see no spec changes in the PR diff.

**Comment by nicolasstucki** [comment]
I meant the addition of the `explain=` tag in the message and how we special case it.

**Comment by nicolasstucki** [comment]
Can I create a copy of `ImplicitNotFound` into the dotty library and update the documentation? Not sure if that could cause conflicts when lading the compiled binaries.

File: library/src/scala/quoted/Quotes.scala [hunk 1]
```diff
@@ -21,7 +22,25 @@ transparent inline def quotes(using q: Quotes): q.type = q
...
+ *  inline def myMacro: Expr[T] =
+ *    ${ /* (quotes: Quotes) ?=> */ myExpr }
+ *  def myExpr(using Quotes): Expr[T] =
+ *    '{ f(${ /* (quotes: Quotes) ?=> */ myOtherExpr }) }
+ *  }
+ *  def myOtherExpr(using Quotes): Expr[U] = '{ ... }
+ *  ```
  */
+
+@implicitNotFound("""explain=Maybe this methods is missing a `(using Quotes)` parameter.
```
Comment (odersky) [inline library/src/scala/quoted/Quotes.scala:37]:
```suggestion
@implicitNotFound("""explain=Maybe this method is missing a `(using Quotes)` parameter.
```

**Review by odersky — COMMENTED** [review]
Small typo, otherwise LGTM

**Comment by sjrd** [comment]
> I meant the addition of the `explain=` tag in the message and how we special case it.

That seems fine to me.

File: tests/neg-custom-args/explain/i16888.check [hunk 2]
```diff
@@ -0,0 +1,14 @@
+-- [E172] Type Error: tests/neg-custom-args/explain/i16888.scala:1:38 --------------------------------------------------
+1 |def test = summon[scala.quoted.Quotes] // error
+  |                                      ^
+  |               No given instance of type quoted.Quotes was found for parameter x of method summon in object Predef
+  |---------------------------------------------------------------------------------------------------------------------
+  | Explanation (enabled by `-explain`)
+  |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+  | Maybe this methods is missing a `(using Quotes)` parameter.
```
Comment (odersky) [inline tests/neg-custom-args/explain/i16888.check:8]:
```suggestion
  | Maybe this method is missing a `(using Quotes)` parameter.
```

**Review by nicolasstucki — APPROVED** [review]

**Pull request merged**